ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala

```diff
@@ -45,74 +73,111 @@ private[backend] object MeteringStorageBackendTemplate {
 
   }
 
+  /** Oracle does not understand true/false so compare against number
+    * @return 0 if the option is unset and non-zero otherwise
+    */
+  def isSet(o: Option[_]): Int = o.fold(0)(_ => 1)
+
+  /** Oracle treats zero length strings as null and then does null comparison
+    * @return 0 if the offset is beforeBegin and non-zero otherwise
+    */
+  def hasBegun(o: Offset): Int = if (o == Offset.beforeBegin) 0 else 1
+
 }
 
-private[backend] class MeteringStorageBackendReadTemplate(ledgerEndCache: LedgerEndCache)
-    extends MeteringStorageReadBackend {
+private[backend] object MeteringStorageBackendReadTemplate extends MeteringStorageReadBackend {
+
+  implicit val offsetToStatement: ToStatement[Offset] =
+    com.daml.platform.store.Conversions.OffsetToStatement
+  implicit val timestampToStatement: ToStatement[Timestamp] =
+    com.daml.platform.store.Conversions.TimestampToStatement
+  implicit val timestampParamMeta: ParameterMetaData[Timestamp] =
+    com.daml.platform.store.Conversions.TimestampParamMeta
 
-  override def transactionMetering(
+  override def reportData(
       from: Time.Timestamp,
       to: Option[Time.Timestamp],
-      applicationId: Option[ApplicationId],
-  )(connection: Connection): Vector[TransactionMetering] = {
+      maybeApplicationId: Option[ApplicationId],
+  )(connection: Connection): ReportData = {
+
+    val ledgerMeteringEnd = assertLedgerMeteringEnd(connection)
+    val participantData = participantMetering(from, to, maybeApplicationId)(connection)
+    val isFinal = to.fold(false)(ledgerMeteringEnd.timestamp >= _)
+    val data = if (isFinal) {
+      participantData
+    } else {
+      val transactionData =
+        transactionMetering(ledgerMeteringEnd.offset, to, maybeApplicationId)(connection)
+      val apps: Set[ApplicationId] = participantData.keySet ++ transactionData.keySet
+      apps.toList.map { a =>
+        a -> (participantData.getOrElse(a, 0L) + transactionData.getOrElse(a, 0L))
+      }.toMap
+    }
+
+    ReportData(data, isFinal)
+
+  }
+
+  private def transactionMetering(
+      from: Offset,
+      to: Option[Time.Timestamp],
```

Comment:
Maybe convey how the interval ends are treated in the parameter names:  `fromIncl/Excl` and `toIncl/Excl`

Reply:
Added some scaladoc to explain the full purpose.